sdk/core/azure-core/test/ut/retry_policy_test.cpp

```diff
@@ -56,23 +56,29 @@ class RetryPolicyTest final : public RetryPolicy {
           std::chrono::milliseconds&,
           double)> shouldRetryOnResponse)
       : RetryPolicy(retryOptions),
-      m_shouldRetryOnTransportFailure(
-          shouldRetryOnTransportFailure != nullptr
-            ? shouldRetryOnTransportFailure
-            : [&](auto options, auto attempt, auto retryAfter, auto jitter) {
-              retryAfter = std::chrono::milliseconds(0);
-              auto ignore = decltype(retryAfter)();
-              return RetryPolicy::ShouldRetryOnTransportFailure(options, attempt, ignore, jitter);
-        }),
+        m_shouldRetryOnTransportFailure(
+            shouldRetryOnTransportFailure != nullptr
+                ? shouldRetryOnTransportFailure
+                : static_cast<decltype(shouldRetryOnTransportFailure)>(
+                    [&](auto options, auto attempt, auto retryAfter, auto jitter) {
```

Comment:
Can you help me understand what the issue was and how this cast helps fix it?

Reply:
GCC5 does not do implicit conversion from a lambda to `std::function`, so it thinks two sides of a ternary operator return different types, which is not allowed. Other thing, on GCC5, lambda `[&]` written within the member initializer, is not smart enough to grab `this` pointer and use it when resolving the function name.

Reply:
So, in the end, the clearest form seems to be to re-initialize in the constructor body. It appears in GCC5, lambdas used within the initializer context are treated as if they are written outside of class context, so they think can't call protected functions.
I could've tried to write a static function that returns the std::function, and use it in the initializer, but I think the current way it is written, should be the most clear to understand.